### ISSUE TITLE
Add "<LangVersion>Latest</LangVersion>" to all projects

### ISSUE THREAD

**Issue by HugoRoss** [exercism/csharp#1561]
Currently, the current C# versions are not supported, one has to use legacy syntax.

Add `<LangVersion>Latest</LangVersion>` to a `<PropertyGroup>` of every project (*.csproj).

**Comment by xtqqczze** [comment]
This shouldn't be necessary, the compiler determines a default language version according to the .NET version, see:
https://docs.microsoft.com/dotnet/csharp/language-reference/configure-language-version

**Comment by HugoRoss** [comment]
Try it out if you don't believe me. Use a simple "if (someParameter is not
null) {}". It does not work without "<LangVersion>Latest</LangVersion>

**Comment by xtqqczze** [comment]
Negated `not` patterns were included in C# 9.0.

@HugoRoss What compiler version are you using? Perhaps you could report the result of putting `#error version` in `exercises\concept\land-grab-in-space\LandGrabInSpaceTests.cs`:

![image](https://user-images.githubusercontent.com/45661989/129195663-2d731ceb-bba3-4b13-b279-fa3dda1b8b5e.png)

**Comment by ErikSchierboom** [comment]
@HugoRoss @xtqqczze You are both right :)

For the current exercises, the C# 9 features can't be used because they target .NET Core 3.1. However, all exercises in this repo have been updated to .NET 5 and when v3 launches, all solutions will be updated to the latest version.

I've just checked that adding an `if (statement is not null)` statement to an exercise works, and it does. So I think we can close this issue. Students will be able to use the C# 9 features once v3 launches in 2 weeks.

**Comment by HugoRoss** [comment]
[image: image.png]

Am Do., 12. Aug. 2021 um 14:24 Uhr schrieb xtqqczze <
***@***.***>:

**Comment by HugoRoss** [comment]
I agree, if the exercises have been upgraded to .NET 5.0 and are released
in a few days we should not bother to change that in the old repo, but it
would still make sense to add it to the new projects as it guarantees that
the most recent language version is always supported (even in the future
when .NET 5.0 produces obsolete warnings and - reading tea leaves - .NET
7.2 / C# 12 are state of the art).

Am Di., 17. Aug. 2021 um 09:01 Uhr schrieb Erik Schierboom <
***@***.***>:

**Comment by ErikSchierboom** [comment]
> but it would still make sense to add it to the new projects as it guarantees that the most recent language version is always supported

We could. It does put the burden a bit more on mentors to always be completely up to date, but 🤷

**Comment by ErikSchierboom** [comment]
From the [official website](https://docs.microsoft.com/nl-nl/dotnet/csharp/language-reference/configure-language-version#c-language-version-reference):

> The compiler accepts syntax from the latest released version of the compiler (including minor version).

Reading that, there is also no harm in adding the property I think. @HugoRoss Would you be willing to submit a PR to add this property to all exercises?

**Comment by xtqqczze** [comment]
> From the [official website](https://docs.microsoft.com/nl-nl/dotnet/csharp/language-reference/configure-language-version#c-language-version-reference):
> 
> > The compiler accepts syntax from the latest released version of the compiler (including minor version).
> 
> Reading that, there is also no harm in adding the property I think. @HugoRoss Would you be willing to submit a PR to add this property to all exercises?

The alternative is to omit the `LangVersion` property, so as to fix the language version to the target framework. The docs you linked specify that for .NET 6.x this will be `10.0`. So in future, you could enable new language features by updating the target framework.

**Comment by ErikSchierboom** [comment]
@mikedamay @robkeim @wolf99 @SleeplessByte Thoughts?

**Comment by HugoRoss** [comment]
Yes, can do.

Am Do., 19. Aug. 2021 um 08:42 Uhr schrieb Erik Schierboom <
***@***.***>:
